### DISCUSSION TITLE
Expose Scanner API for C

### DISCUSSION THREAD

**Discussion by prashanthellina** [duckdb/duckdb#14261] [Ideas]
I'm thinking about implementing a DuckDB scanner abstraction for a custom file format however my code is in C. I don't see the C-API doc list any scanner related API methods. Can this be exposed please?

**Comment by Tishj** [comment]
Which C++ API methods are you referring to that you want to have exposed?

**Reply by prashanthellina** [reply]
I don't know specifically which methods but I'm thinking of whatever kinds of constructs are used by https://github.com/duckdb/postgres_scanner .

**Comment by Maxxen** [comment]
Is the [table function](https://github.com/duckdb/duckdb/blob/123b82b9053c4843559035b6723c867b2618b2d9/src/include/duckdb.h#L3087) API not sufficient?

**Reply by prashanthellina** [reply]
If my understanding is correct, table functions don't have a provision for "filter pushdown".  They appear to support only "projection pushdown".

**Reply by Tishj** [reply]
https://github.com/duckdb/duckdb/blob/5a9a382a573b107a38f5ee277619b362d5079c32/src/include/duckdb/function/table_function.hpp#L300

Nope, filter pushdown is supported

**Reply by prashanthellina** [reply]
Oh! That is interesting. I looked at this documentation page for the C-API and could not find any function for setting that filter pushdown is supported for a table function (https://duckdb.org/docs/api/c/table_functions). There is a function for setting support for projection pushdown though (https://duckdb.org/docs/api/c/table_functions#duckdb_table_function_supports_projection_pushdown). Am I missing something?

**Reply by prashanthellina** [reply]
@Tishj @Maxxen Is it possible to use table function's filter pushdown functionality through the C-API?

**Reply by Maxxen** [reply]
Doesn't seem like it. However, filter pushdown is an optimization, not a requirement, I'd advice you to start working on whatever scanner extension you got planned and then have a look at extending the c-api. At that point you might even be familiar with working with DuckDB enough that you could implement it yourself - happy to review a PR!

**Reply by prashanthellina** [reply]
Thank you @Maxxen

**Reply by prashanthellina** [reply]
@Maxxen I was able to successfully create a table function using the C-API. I want to try to expose filter pushdown functionality in the C-API. Is there any example of creating a table function in C++ (preferably a simple one and one that uses filter pushdown)?

My thought process is that if I can write a table function in C++ that uses filter pushdown then I will gain enough knowledge to attempt exposing via C-API.

**Reply by prashanthellina** [reply]
@Maxxen @Tishj I've issued a PR (https://github.com/duckdb/duckdb/pull/14591) for this. Can you please check it out and provide feedback? Thank you.
